File: src/Engines/MediumEngine.hs
module Engines.MediumEngine (mediumEngine) where
import Engines.RandomEngine

makeMove :: Int -> [Int] -> (Int, Int)
makeMove _ [] = (-1,-1)
makeMove i (x : xs) = if x `mod` 4 /= 0 then (x `mod` 4, i) else makeMove (i + 1) xs  

mediumEngine :: [Int] -> (Int, Int)
mediumEngine xs =
    let res = makeMove 0 xs
    in if res == (-1, -1) then randomEngine 0 xs else res

File: prolog/src/Engines/MediumEngine.pl
:- include('RandomEngine.pl').

haveMove([]) :- false.
haveMove([X|_]) :- (X mod 4) =\= 0, !.
haveMove([_|Y]) :- haveMove(Y).

makeMove(Idx,[X|_], [V, Idx]) :-
    V is X mod 4,
    V =\= 0,
    !.

makeMove(Idx, [_|Y], R) :- 
    Idx2 is Idx + 1,
    makeMove(Idx2, Y, R).

mediumEngine(Stacks, R) :- 
    haveMove(Stacks),
    makeMove(0, Stacks, R),!.

mediumEngine(Stacks, R) :- randomEngine(Stacks, R).
